### PR TITLE
Redesign of std.uni matcher API

### PR DESCRIPTION
**Pull request by DmitryOlshansky**
This continues the work of pull #3710 and unifies API for all of decode-less matchers.

In particular:
 - any functor on a RA-range of code units => UtfLookup is an UtfMatcher
 - all of match/test/skip helpers are available for "free" with UFCS
 - all of matchers are nothrow for speed, inspect UtfLookup result directly to handle broken encodings and more

This may need a bit polish - please review carefully.

### CODE REVIEW

File: std/uni.d [hunk 1]
```diff
@@ -4517,18 +4524,20 @@ public struct MatcherConcept
...
-        assert(!m.subMatcher!(2,3,4).test(square)); // ASCII no covered
+        assert(!m.subMatcher!(2,3,4).test(square)); // ASCII not covered
         assert(m.subMatcher!1.match(square)); // ASCII-only, works
         assert(!m.subMatcher!1.test(square)); // unicode '²'
         assert(m.subMatcher!(2,3,4).match(square));  //
         assert(square == "");
         wstring wsquare = "2²";
         auto m16 = utfMatcher!wchar(unicode.Number);
+        /*assert(m16.match(wsquare)); // not matched
+        assert(m16.match(wsquare));*/
```
Comment (Hackerpilot) [inline std/uni.d]: Adding commented-out code?

File: std/uni.d [hunk 2]
```diff
@@ -5134,6 +4983,708 @@ public auto utfMatcher(Char, Set)(Set set) @trusted
...
+// helper for assert/debug prints
+string utfWordAsBinary()(uint val, size_t sz=0)
+{
+    import std.algorithm : reverse;
+    import core.bitop;
+    import std.format;
+    char[] res;
+    if(sz == 0)
+        sz = (bsr(val)+8)/8;
+    while(sz){
```
Comment (Hackerpilot) [inline std/uni.d]: Brace should go on next line.

**Comment by DmitryOlshansky** [comment]
@Hackerpilot fixed, also a couple of instance of bad style in the source

**Comment by Hackerpilot** [comment]
#3710 should be reviewed before this pull, correct?

**Comment by DmitryOlshansky** [comment]
@Hackerpilot yes #3710 must be reviewed/merged first

**Comment by DmitryOlshansky** [comment]
Please review but do not hurry with a merge I'm investigating performance issues and looking at codegen.

**Comment by DmitryOlshansky** [comment]
I admit defeat. Exactly like the last time - trying to get a universal function that returns both length and test flags is futile, compilers simply fail to optimize it properly. I would have to rework #3710 to follow the old approach of having 3 functions that may throw on error...

Proper compiler e.g. LDC inlines the hell out of decode making it real fast. Then the lookup for Trie of dchars is unrolled to 10-12 instructions (how beautiful!) and the resulting combo is hard to surpass.

Now getting back to universal matcher - it can't inline its body into the scope of test hence have to do a lot of bit-packing and what not. All in all the competition here is doing less calls and pointless push/pop of registers. I can't figure out a way to micro-optimize further.

**Pull request closed without merging**